icicle/backend/cpu/include/ntt_tasks.h

```diff
@@ -0,0 +1,1206 @@
+    int find_or_generate_coset(std::unique_ptr<S[]>& arbitrary_coset);
+    void h1_reorder(E* elements);
+    eIcicleError
+    reorder_and_refactor_if_needed(E* elements, NttTaskCordinates ntt_task_cordinates, bool is_top_hirarchy);
+    eIcicleError
+    hierarchy1_push_tasks(E* input, NttTaskCordinates ntt_task_cordinates, NttTasksManager<S, E>& ntt_tasks_manager);
+    eIcicleError h0_cpu_ntt(E* input, NttTaskCordinates ntt_task_cordinates);
+    eIcicleError handle_pushed_tasks(
+      TasksManager<NttTask<S, E>>* tasks_manager, NttTasksManager<S, E>& ntt_tasks_manager, int h1_layer_idx);
+
+  private:
+    uint64_t bit_reverse(uint64_t n, int logn);
+    uint64_t idx_in_mem(NttTaskCordinates ntt_task_cordinates, int element);
+    void refactor_output_h0(E* elements, NttTaskCordinates ntt_task_cordinates);
+  }; // class NttCpu
+
+  /**
+   * @brief Manages task dependency counters for NTT computation, tracking readiness of tasks to execute.
+   *
+   * This class tracks and manages counters for tasks within the NTT hierarchy, determining when tasks are ready to
+   * execute based on the completion of their dependencies.
+   *
+   * @param h1_layer_idx Index of the h1 layer this counter set belongs to.
+   * @param nof_h0_layers Number of h0 layers in the current h1 layer.
+   * @param nof_pointing_to_counter Number of counters pointing to each h0 layer.
+   * @param h0_counters A 3D vector of shared pointers to counters for each sub-NTT in h0 layers.
+   * @param h1_counters A vector of shared pointers to counters for each sub-NTT in h1 layers, used to signal when an
+   * h1_subntt is ready for reordering.
+   *
+   * @method TasksDependenciesCounters(NttSubLogn ntt_sub_logn, int h1_layer_idx) Constructor that initializes the
+   * counters based on NTT structure.
+   * @method bool decrement_counter(NttTaskCordinates ntt_task_cordinates) Decrements the counter for a given task and
+   * returns true if the task is ready to execute.
+   * @method int get_nof_pointing_to_counter(int h0_layer_idx) Returns the number of counters pointing to the given h0
+   * layer.
+   * @method int get_nof_h0_layers() Returns the number of h0 layers in the current h1 layer.
+   */
+  class TasksDependenciesCounters
+  {
+  public:
+    // Constructor that initializes the counters
+    TasksDependenciesCounters(NttSubLogn ntt_sub_logn, int h1_layer_idx);
+
+    // Function to decrement the counter for a given task and check if it is ready to execute. if so, return true
+    bool decrement_counter(NttTaskCordinates ntt_task_cordinates);
+    int get_nof_pointing_to_counter(int h0_layer_idx) { return nof_pointing_to_counter[h0_layer_idx]; }
+    int get_nof_h0_layers() { return nof_h0_layers; }
+
+  private:
+    int h1_layer_idx;
+    int nof_h0_layers;
+    std::vector<int> nof_pointing_to_counter; // Number of counters for each layer
+
+    // Each h1_subntt has its own set of counters
+    std::vector<std::vector<std::vector<std::shared_ptr<int>>>>
```

Comment:
I don't think that you need a shared ptr here. simply int.
Also this 3d cube is full? or is it sparse? I don't know how efficient is all those redirections if you use it a lot. Maybe you don't.